File: src/test/java/io/icednut/algorithm/exercise/bit/leet2220/SolutionTest.java
package io.icednut.algorithm.exercise.bit.leet2220;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class SolutionTest {

    @Test
    void test01() {
        Solution solution = new Solution();

        // 1010 -> 1011
        // 1011 -> 1111
        // 1111 -> 0111
        int output = solution.minBitFlips(10, 7);

        assertEquals(3, output);
    }

    @Test
    void test02() {
        Solution solution = new Solution();

        // 011 -> 010
        // 010 -> 000
        // 000 -> 100
        int output = solution.minBitFlips(3, 4);

        assertEquals(3, output);
    }

    @Test
    void test03() {
        Solution solution = new Solution();

        // 011 -> 010
        // 010 -> 000
        // 000 -> 100
        int output = solution.minBitFlips(35, 22);

        assertEquals(4, output);
    }

    @Test
    void test04() {
        Solution solution = new Solution();

        int output = solution.minBitFlips(0, 1);

        assertEquals(1, output);
    }
}

File: src/test/kotlin/io/icednut/algorithm/exercise/graph/leet785/SolutionTest.kt
package io.icednut.algorithm.exercise.graph.leet785

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

/**
 * TODO
 * - BFS로 순회
 * - Queue에 좌표와 color를 함께 넣는다.
 * - color에 해당하는 hashSet에 인덱스를 넣는다.
 * - red color와 green color의 길이가 같으면 bipatite graph, 아니면 common graph
 */
class SolutionTest {

    @Test
    fun test01() {
        val solution = Solution()

        val result = solution.isBipartite(
            arrayOf(
                intArrayOf(1, 2, 3),
                intArrayOf(0, 2),
                intArrayOf(0, 1, 3),
                intArrayOf(0, 2),
            )
        )

        assertEquals(false, result)
    }

    @Test
    fun test02() {
        val solution = Solution()

        val result = solution.isBipartite(
            arrayOf(
                intArrayOf(1, 3),
                intArrayOf(0, 2),
                intArrayOf(1, 3),
                intArrayOf(0, 2),
            )
        )

        assertEquals(true, result)
    }

    @Test
    fun test03() {
        val solution = Solution()

        val result = solution.isBipartite(
            arrayOf(
                intArrayOf(3),
                intArrayOf(2, 4),
                intArrayOf(1),
                intArrayOf(0, 4),
                intArrayOf(1, 3),
            )
        )

        assertEquals(true, result)
    }

    @Test
    fun test04() {
        val solution = Solution()

        val result = solution.isBipartite(
            arrayOf(
                intArrayOf(4),
                intArrayOf(),
                intArrayOf(4),
                intArrayOf(4),
                intArrayOf(0, 2, 3),
            )
        )

        assertEquals(true, result)
    }

    @Test
    fun test05() {
        val solution = Solution()

        val result = solution.isBipartite(
            arrayOf(
                intArrayOf(),
                intArrayOf(3),
                intArrayOf(),
                intArrayOf(1),
                intArrayOf(),
            )
        )

        assertEquals(true, result)
    }

    @Test
    fun test06() {
        val solution = Solution()

        val result = solution.isBipartite(
            arrayOf(
                intArrayOf()
            )
        )

        assertEquals(true, result)
    }
}